src/core/proof.ml

```diff
@@ -22,6 +22,9 @@ module Goal :
 
     (** [simpl g] simplifies the given goal, evaluating its type to SNF. *)
     val simpl : t -> t
+
+    (** Comparison function. *)
+    val compare : t -> t -> int
```

Comment:
I would replace that by:
```ocaml
module Set : Set.S with type elt = t
```
And the corresponding implementation:
```ocaml
module Set = Set.Make(
  type struct
    type t = t (* you might need to do things differently here, I am not sure if this is valid. *)
    let compare {goal_meta = m1; _} {goal_meta = m2; _} = m1.key - m2.key
  end)
```

You can then use sets of goals instead of sorted lists of goals.

Reply:
And again, there is no order on goals, so we really do not want a `compare` function to be exposed.

Reply:
In the end, we need lists, not sets, as the order is important (new goals should come first).

Reply:
I still don't like to expose a compare function here.